status/status.go

```diff
@@ -50,78 +50,56 @@ import (
 	"google.golang.org/grpc/codes"
 )
 
-// Status provides access to grpc status details and is implemented by all
-// errors returned from this package except nil errors, which are not typed.
-// Note: gRPC users should not implement their own Statuses.  Custom data may
-// be attached to the spb.Status proto's Details field.
-type Status interface {
-	// Code returns the status code.
-	Code() codes.Code
-	// Message returns the status message.
-	Message() string
-	// Proto returns a copy of the status in proto form.
-	Proto() *spb.Status
-	// Err returns an error representing the status.
-	Err() error
-}
-
-// okStatus is a Status whose Code method returns codes.OK, but does not
-// implement error.  To represent an OK code as an error, use an untyped nil.
-type okStatus struct{}
-
-func (okStatus) Code() codes.Code {
-	return codes.OK
-}
-
-func (okStatus) Message() string {
-	return ""
-}
-
-func (okStatus) Proto() *spb.Status {
-	return nil
-}
+// statusError is an alias of a status proto.  It implements error and Status,
+// and a nil statusError should never be returned by this package.
+type statusError spb.Status
 
-func (okStatus) Err() error {
-	return nil
+func (se *statusError) Error() string {
+	p := (*spb.Status)(se)
+	return fmt.Sprintf("rpc error: code = %s desc = %s", codes.Code(p.GetCode()), p.GetMessage())
 }
 
-// statusError contains a status proto.  It is embedded and not aliased to
-// allow for accessor functions of the same name.  It implements error and
-// Status, and a nil statusError should never be returned by this package.
-type statusError struct {
-	*spb.Status
+func (se *statusError) status() *Status {
+	return &Status{s: (*spb.Status)(se)}
 }
 
-func (se *statusError) Error() string {
-	return fmt.Sprintf("rpc error: code = %s desc = %s", se.Code(), se.Message())
+// Status represents an RPC status code, message, and details.  It should be
+// created with New, Newf, or FromProto.
+type Status struct {
+	s *spb.Status
 }
 
-func (se *statusError) Code() codes.Code {
-	return codes.Code(se.Status.Code)
+// Code returns the status code contained in s.
+func (s *Status) Code() codes.Code {
+	return codes.Code(s.s.Code)
 }
 
-func (se *statusError) Message() string {
-	return se.Status.Message
+// Message returns the message contained in s.
+func (s *Status) Message() string {
+	return s.s.Message
 }
 
-func (se *statusError) Proto() *spb.Status {
-	return proto.Clone(se.Status).(*spb.Status)
+// Proto returns s's status as an spb.Status proto message.
+func (s *Status) Proto() *spb.Status {
+	return s.s
 }
 
-func (se *statusError) Err() error {
-	return se
+// Err returns an immutable error representing s; returns nil if s.Code() is
+// OK.
+func (s *Status) Err() error {
+	if s.Code() == codes.OK {
+		return nil
+	}
+	return (*statusError)(proto.Clone(s.s).(*spb.Status))
```

Comment:
To make it impossible to mutate the statusError and change the Code into OK.

Without this, but the rest of the code as-is, one could do:

status := status.New(codes.Internal, "blah")
p := status.Proto()
err := status.Err()
p.Code = int32(codes.OK)

Maybe status.Proto() and FromProto() should do the clone instead, making Status immutable, which is how I had it initially.